abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py

```diff
@@ -156,3 +146,137 @@ def signaling(
+
+    # 1. Calculating closed angle between object and agent according to the
+    # position of the object.
+    # 2. Calculating visual projection size according to visual angle on
+    # the agent's retina according to distance between agent and object
+    for i, obj_position in enumerate(object_positions):
+        # continue if the object and the agent position completely coincide
+        if obj_position[0] == position[0] and obj_position[1] == position[1]:
+            continue
+
+        # center of obstacle (as it must be another agent)
+        object_center = obj_position + radius
+
+        # vector between agent center and object center
+        v2 = object_center - agents_center
+
+        # calculating closed angle between v1 and v2
+        # (rotated with the orientation of the agent as it is relative)
+        closed_angle = angle_between(v1, v2)
+        closed_angle = (closed_angle % (2 * np.pi))
+        # at this point closed angle between 0 and 2pi, but we need it between
+        # -pi and pi
+        # we also need to take our orientation convention into consideration to
+        # recalculate theta=0 is pointing to the right
+        if 0 < closed_angle < np.pi:
+            closed_angle = -closed_angle
+        else:
+            closed_angle = 2 * np.pi - closed_angle
+
+        distance = np.linalg.norm(object_center - agents_center)
+
+        # calculating the visual angle from focal agent to target
+        vis_angle = 2 * np.arctan(radius / (1 * distance))
+
+        # finding where in the retina the projection belongs to
+        phi_target = find_nearest(phis, closed_angle)
+
+        # if target is visible we save its projection into the VPF
+        # source data
+        if fov[0] <= closed_angle <= fov[1]:
+            # the projection size is proportional to the visual angle.
+            # If the projection is maximal (i.e. taking each pixel of the
+            # retina) the angle is 2pi from this we just calculate the
+            # projection size using a single proportion
+            proj_size = (vis_angle / (2 * np.pi)) * v_field_resolution
+            proj_start = int(phi_target - np.floor(proj_size / 2))
+            proj_end = int(phi_target + np.floor(proj_size / 2))
+
+            # circular boundaries to the VPF as there is 360 degree vision
+            if proj_start < 0:
+                v_field[i, v_field_resolution + proj_start:v_field_resolution] = 1
+                proj_start = 0
+            if proj_end >= v_field_resolution:
+                v_field[i, 0:proj_end - (v_field_resolution - 1)] = 1
+                proj_end = v_field_resolution
+
+            v_field[i, proj_start:proj_end] = 1
+
+            if object_meters is not None:
+                v_field[i] *= object_meters[i]
+
+    # post_processing and limiting FOV
+    # flip field data along second dimension
+    # TODO: why we need to flip the field?
+    # v_field_post = np.flip(v_field, axis=1)
```

Comment:
@mezdahun I'm not sure why the field of view was flipped in the previous version. It seems to work correctly without flipping 🤔

Reply:
Originally we flipped the field of view to match with the left/right convention. (i.e. what is on the right side of the retina is from the right side of the agent and vica versa). I don't know in the refactored implementation why you don't need to flip this.

Reply:
That was a bug fixed below 😓 See db2331123873f14b12371b043e18cc3b3963ab01 and 6da1095f7865d0381ea0465ee32762df398ae3c5